LRA-415-art-survey-add-users-roles-policy

rsmoke opened the pull request:
This is the whole tuna for authorization. 
There are checks for what access a user has and what is displayed based on department roles.

- When a user logs in they are authenticated via Shibboleth
- Upon success they have session variable set that is labelled :uniqname
- There are various methods in application_controller that will take the user's uniqname session variable return appropriate data items depending on the authorization that is required on a controller or a variable list.
- The controllers that govern Role and Permission are specifically locked down to SuperUser status
- Access controller can be a used by SuperUser or Department Administrator only
- ArtItem controller is accessible by anyone who has an access in the system
- The StaticPage#home is not protected and allows public viewing of the home page and it's message

The ROLE / PERMISSION / ACCESS bits

- A role is set by a SuperUser to describe currently one of three types

> SuperUser - see-er / editor /deleter of all things

> Department Administrator - see-er / editor /deleter of department specific related items

> Recorder - See-er / editor of department specific related items

- A permission is matching a role to a department ie. Dean's - Recorder
- Accesses are matching a uniqname to a permission

britaumich reviewed (commented):
Since the system is s-o-o-o-o-o-o-o (too) complicated, it would be nice if you add some description. That would help to test the process.
Please describe the authorization process.

rsmoke commented:
> Since the system is s-o-o-o-o-o-o-o (too) complicated, it would be nice if you add some description. That would help to test the process. Please describe the authorization process.

See edited PR 'initial' comment

britaumich commented:
Let's start from the beginning. On my localhost, the roles table has three roles, permissions and accesses tables are empty.
I can login; I see no links. 

But I can go directly to http://localhost:3000/art_items - I see no record, but the button [New art item] is available. I can create a record for any department. I can see the record.

I can access and update:

- http://localhost:3000/page_informations)

The following URLs give different errors:

- http://localhost:3000/annotations, 
- http://localhost:3000/appraisal_types
- http://localhost:3000/departments

Not authorized (good!):

- http://localhost:3000/accesses
- http://localhost:3000/permissions
- http://localhost:3000/roles

rsmoke commented:
> Can access and update without a specific permission
> * http://localhost:3000/page_informations)
> 
> The following URLs give different errors:
> 
> * http://localhost:3000/annotations,
> * http://localhost:3000/appraisal_types
> * http://localhost:3000/departments

I updated the before actions to the ones that actually will do a proper √ for access approval.

britaumich commented:
How can I make my account a superuser? In rails console?
Right now without any permissions I can create a new art_item and see show pages.

britaumich commented:
I created Access records for my account.
1. If I go to http://localhost:3000/accesses I see my account and I can create a new record. But in both cases, I get a flash message 'Not Authorized."
The same is true for http://localhost:3000/permissions.
Is the super_user_department_admin_access_authorized! method correct? It displays 'Not Authorized' when it gives access or denies access. I created an um171712 account with recorder permission and for this account, authorization works correctly.
2. seed.rb file has errors (both syntax and logical). Do you want me to fix them? Or describe them here?

rsmoke commented:
> I created Access records for my account.
> 
> 1. If I go to http://localhost:3000/accesses I see my account and I can create a new record. But in both cases, I get a flash message 'Not Authorized."
>    The same is true for http://localhost:3000/permissions.
>    Is the super_user_department_admin_access_authorized! method correct? It displays 'Not Authorized' when it gives access or denies access. I created an um171712 account with recorder permission and for this account, authorization works correctly.
> 2. seed.rb file has errors (both syntax and logical). Do you want me to fix them? Or describe them here?

Fixed the notification message and updated the seeds.rb

britaumich reviewed (commented):
In the seed file, all these tables need updated_by fields to have data, otherwise, show_user(id) gives an error
```
def show_user(id)
    User.find(id).updated_by_name
 end 
```

rsmoke commented:
> In the Department controller in the create method, when a new department is created permissions for this department should be created at the same time, I think.

Not necessarily

britaumich commented:
> Not necessarily

Of course, not, but then after creating a department you have to create permissions for that department. Well, you'll take care of that in the admin interface ... which will be as complicated as the ROLE / PERMISSION / ACCESS process.

britaumich reviewed (changes requested):
As a user with no access records, I still can go to http://localhost:3000/art_items, see [New art item] button, and go directly to http://localhost:3000/art_items/11

This has been fixed

britaumich commented:
> Please share your table structure so I can see what you have configured locally

```
Role.all.pluck(:id, :title)
  Role Pluck (0.5ms)  SELECT "roles"."id", "roles"."title" FROM "roles"
=> [[4, "SuperUser"], [5, "Department Administrator"], [6, "Recorder"]] 
```

```
Permission.all.pluck(:role_id, :department_id)
  Permission Pluck (0.5ms)  SELECT "permissions"."role_id", "permissions"."department_id" FROM "permissions"
=> [[4, 26], [6, 26]]  
```

```
Access.all.pluck(:permission_id, :uniqname)
  Access Pluck (0.6ms)  SELECT "accesses"."permission_id", "accesses"."uniqname" FROM "accesses"
=> [[1, "brita"], [1, "rsmoke"]]  
```

```
User.all.pluck(:id, :uniqname)
  User Pluck (0.4ms)  SELECT "users"."id", "users"."uniqname" FROM "users"
=> [[2, "brita"], [3, "um171712"]]   
```

User um171712 can login and go to http://localhost:3000/art_items, see [New art item] button, and go directly to http://localhost:3000/art_items/11

rsmoke commented:
> User um171712 can login and go to http://localhost:3000/art_items, see [New art item] button, and go directly to http://localhost:3000/art_items/11

I was expecting the result of the before_action :access_authorized! to be the equivalent of:
```
    if access_authorized!
      @new_annotation = Annotation.new(art_item: @art_item)
      @annotations = @art_item.annotations.order("created_at DESC")
    else
      redirect_to root_path
      flash.alert = "Not Authorized."
    end
```
Here is a good example where that did not work and a test would have captured the problem

rsmoke commented:
@britaumich I refactored the access over the art_item controller to prevent unauthorized miscreants

anantaUM commented on app/controllers/accesses_controller.rb:
```diff
@@ -64,6 +66,10 @@ def set_access
       @access = Access.find(params[:id])
     end
 
+    def get_accesses_collection
```
Consider moving this method inside ApplicationController along with other helper methods.

britaumich replied:
@anantaUM  get_accesses_collection method is a private method for the AccessesController. I think It should not be available to other controllers.

anantaUM replied:
@britaumich - ApplicationController has all helper methods related to Role / Permission / Access. Hence, I think it is better to move it there to avoid repetition in case we want to use same logic elsewhere in future.

britaumich replied:
@anantaUM ApplictionController has methods that are available to the whole application. If a method is called only in one controller it should stay in that controller. 
get_accesses_collection is called only once, and it should not be a method at all.  
Moving part of this statement 
`@accesses = Access.where(permission_id: current_user_permissions).order(:uniqname)
` 
to a method complicates things and adds confusion.

rsmoke replied:
moved to Application Controller

anantaUM commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,64 @@
+  helper_method :role_object
+
+  def is_user_a?( role_title )
+    if Permission.where(role_id: role_object( role_title )).exists?
```
Keep it DRY. 

_Permission.where(role_id: role_object( role_title ))_ is repeated on line 10 and 11. Please store it in a variable and then use it in both places.

rsmoke replied:
Dry is the right way to go

anantaUM commented on app/controllers/art_items_controller.rb:
```diff
@@ -96,13 +93,22 @@ def set_art_item
+      ArtItem.active_with_departments.where(department_id: current_user_departments)
+    end
+     
+    def check_for_authorized_access
```
Consider moving this method inside ApplicationController along with other helper methods.

rsmoke replied:
I did move it

anantaUM commented on app/helpers/application_helper.rb:
```diff
@@ -37,4 +37,26 @@ def render_flash_stream
+
+  def department_admins( department )
+    #collection of a department's admins
+    deptadmin_permissions = Permission.where(role_id: department_admin_role_object, department_id: department)
```
Where is _department_admin_role_object_ defined ?

anantaUM commented on app/helpers/application_helper.rb:
```diff
@@ -37,4 +37,26 @@ def render_flash_stream
+
+  def department_recorders( department = nil )
+    #collection of a department's recorders
+    recorder_permissions = Permission.where(role_id: recorder_role_object, department_id: department)
```
Where is _recorder_role_object_ defined ?

rsmoke replied:
This was removed and the helper here is a vestige that also is now removed

anantaUM reviewed (commented):
@rsmoke - Added review comments.

anantaUM commented:
General question:

@rsmoke  - Since "Super User" is not tied to any department, I think we might have to set some dummy department id value for its entry in the Permissions table. But this opens up the possibility of have multiple "Super User" entries in the permissions table which we don't want (  Since Permissions table is acting as a reference table ). Is there a way to limit the "Super User" entry in the Permissions table to 1 ?

britaumich commented:
I know you did not ask for my review again, but since I get emails I tested the branch.
If I log in as a recorder (or a super user) and try to create a new item, I get the following error (is @departments_list correct?)

![Screen Shot 2022-12-12 at 5 40 24 PM](https://user-images.githubusercontent.com/56089936/207170940-15bd303b-ec0b-467e-9322-ef09499581d3.png)

britaumich commented:
To fix the @department_list error,
In the art_items_controller replace the set_departments_list method (lines 95-97) with
```
def set_departments_list
      @departments_list = current_user_departments
end
```

rsmoke commented:
> By the way, why do we need to reassign current_user_departments to another variable? Could you use current_user_departments in the art_item form instead of @departments_list? It works.

This is true but I like having the instance variable in the form

britaumich commented:
> This is true but I like having the instance variable in the form

I agree, but then maybe you don't need [helper_method :current_user_departments] in the application controller which makes the current_user_departments method available in views.

britaumich commented:
I tested what is allowed for Recorders and Department Administrators. They both have access only to their departments, nothing else. They both have access to the archive button which is doing nothing right now and there is no delete button. We can define permissions to delete/archive later when we create the delete/archive process.

rsmoke commented:
> Similar to _dependent: :destroy_ for _:accesses_ in _permission_ model, I think we might need it for _department_ and _role_ model as well. For example - If a certain department is removed, then we want all its permissions and accesses to be removed. Your thoughts ?

@anantaUM I added the dependent: :destroy on the has_many relationships. I also refactored the authorization to DRY it up and for readability

britaumich commented:
Everything seems to be working (different roles and permissions).
I don't think I like 
```
class Department < ApplicationRecord
  has_many :art_items, dependent: :destroy
```
We still don't know what we are doing with deleting/archiving art items. But if an art_item is in the system it should not be deleted with a department automatically. It might be moved to a new owner or archived or something else. But not deleted.

rsmoke reviewed (commented):
done

britaumich reviewed (approved):
Let's merge this 'beast' (strikethrough) beauty and solve problems (if there will be any) as they arise or if they arise.

rsmoke commented:
Ananta wanted to smack his peepers on it one more time

rsmoke commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,95 @@
+    Access.where(uniqname: session[:user_uniqname])
+  end
+
+  def current_department_admin_user_department_ids
```
This suggestion does not catch all the required bits for the Department Administrator cases.
The catch here is that 
- a SuperUser sees everything -- easy call to MODEL.all _which your refactor suggestion works well for this case_
- a Department Administrator see only there departments but they see 'all the associated records' within "their" department. This means I need the current_user (a dept admin in this case) AND the records that are associated with their assigned dept(s) (the concat on line 80)
- recorder see just the their records _which your refactor suggestion works well for this case_

rsmoke replied:
After further discussion with @anantaUM his refactor was correct and the overall logic was skewed. I added his tweaks to the code.
I am now fixing an accesses_controller misconfiguration and one last update to this PR should make the best ever authorization system

britaumich reviewed (approved):
I'll approve it again

anantaUM commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,92 @@
+    elsif is_department_admin_user!
+      Department.where(id: current_department_admin_user_department_permissions_collection.pluck(:department_id))
+    else
+      Department.where(id: current_user_permissions.pluck(:department_id)).uniq
```
@rsmoke - We can remove _uniq_ from this query as well since a user cannot have multiple roles for the same department. So, list of departments will always be unique.

rsmoke replied:
I removed the .uniq
Thanks for catching that

anantaUM reviewed (approved):
Approved
